Fix method stub creation for function attributes

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37994 Fix method stub creation for function attributes**

Before, reassigning a method in a module (like `forward = _forward`)
didn't work, because we look at the function object's name for our def
name when building AST. Mkae that overrideable to handle cases like
reassignment

Differential Revision: [D21444535](https://our.internmc.facebook.com/intern/diff/D21444535)

suo committed:
Fix method stub creation for function attributes

Before, reassigning a method in a module (like `forward = _forward`)
didn't work, because we look at the function object's name for our def
name when building AST. Mkae that overrideable to handle cases like
reassignment

[ghstack-poisoned]

zdevito commented on torch/jit/frontend.py:
```diff
@@ -145,7 +145,7 @@ def get_jit_class_def(cls, self_name):
 
 
-def get_jit_def(fn, self_name=None):
+def get_jit_def(fn, self_name=None, def_name_override=None):
```
def_name_override should probably not be optional or an override, but actually just the name to use everytime. It seems like otherwise, it will cause bugs when people don't realize the same thing this bug addresses: that this is the name of the field of this method. For instance frontend.py:136 is probably wrong too but wasn't caught in this change.

zdevito reviewed (approved):
Sounds good, one potential bug

suo committed:
Update on "Fix method stub creation for function attributes"


Before, reassigning a method in a module (like `forward = _forward`)
didn't work, because we look at the function object's name for our def
name when building AST. Mkae that overrideable to handle cases like
reassignment

Differential Revision: [D21444535](https://our.internmc.facebook.com/intern/diff/D21444535)

[ghstack-poisoned]
